### PR TITLE
Suggest comma when writing `println!("{}" a);`

### LINKED ISSUE #49370: Improve error message: No rules expected the token 'a'
Let's say you are a beginner to Rust and you forget a comma while using `println!`. (This could happen to anyone, but let's specifically imagine that you do not know the language and its features well.)

```rust
fn main() {
    let a = 1;
    println!("{}" a);
}
```

Here is the error you would get:
```
error: no rules expected the token `a`
 --> src/main.rs:3:19
  |
3 |     println!("{}" a);
  |                   ^

error: aborting due to previous error

error: Could not compile `playground`.
```

Since you're a beginner, it is likely that macros are going to be one of the last features you touch. So "no rules expected the token" would not make any sense to you.

Now let's look at what happens if you forget a comma in a regular function:

```rust
fn main() {
    let a = 1;
    foo("{}" a);
}

fn foo(q: &str, x: i32) {
}
```

Here is the error that you get:
```
error: expected one of `,`, `.`, `?`, or an operator, found `a`
 --> src/main.rs:3:14
  |
3 |     foo("{}" a);
  |              ^ expected one of `,`, `.`, `?`, or an operator here

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

This is much better because it makes it clear that we forgot one of the options presented and it is pretty obvious how to fix it.

Since we know exactly which tokens a macro expects, wouldn't it be possible to generate a similar looking (or different but informative) message for macros too? I think this would really enhance the experience especially for beginners who may not understand "no rules expected the token" errors in macros.

Please pardon me if this is a duplicate. I reported this because when I encountered it, I was confused for a second. If I can get confused by this error message after 3 years of programming in Rust, a beginner is certainly likely to be stumped for much longer. It would be awesome if this could improve! :smile:

### PR DESCRIPTION
**Pull request by estebank**
Fix #49370.

### CODE REVIEW

File: src/libsyntax/tokenstream.rs [hunk 1]
```diff
@@ -182,6 +182,31 @@ pub struct TokenStream {
     kind: TokenStreamKind,
 }
 
+impl TokenStream {
+    /// Given a `TokenStream` with a `Stream` of only two arguments, return a new `TokenStream`
+    /// separating the two arguments with a comma for diagnostic suggestions.
+    pub(crate) fn add_comma(&self) -> Option<(TokenStream, Span)> {
+        // Used ot suggest if a user writes `println!("{}" a);`
```
Comment (oli-obk) [inline src/libsyntax/tokenstream.rs]: ot -> to

File: src/libsyntax/tokenstream.rs [hunk 2]
```diff
@@ -182,6 +182,31 @@ pub struct TokenStream {
...
+        if let TokenStreamKind::Stream(ref slice) = self.kind {
+            if slice.len() == 2 {
```
Comment (oli-obk) [inline src/libsyntax/tokenstream.rs]: So this doesn't work for `println!("{} {}" a, b);` or `println!("{} {}", a b);`?

If it does work, please add tests, if it does not work, would it be a lot of effort to fix that?

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by oli-obk** [comment]
r? @oli-obk

Advanced mode: Can we ask the macro matcher how far each arm got and suggest the next possible matches in the spirit of "expected one of `foo`, `bar`, `boo`, found `a`"?

File: src/libsyntax/tokenstream.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/tokenstream.rs]: The further we go from the `"{}" a` case, the less likely we're to find a suitable fix by just checking against comma separated list. Having said that, making sure that every element is separated by commas should be easy to add.

**Comment by estebank** [comment]
> Can we ask the macro matcher how far each arm got and suggest the next possible matches in the spirit of "expected one of foo, bar, boo, found a"?

@oli-obk that sounds interesting to do... I'll look into it this week.

**Comment by pietroalbini** [comment]
Ping from triage @estebank! It's been a while since we heard from you, will you have time to work on this again?

**Comment by estebank** [comment]
@pietroalbini will do soonish.

@oli-obk I tried to get this for arbitrary sequences (finding a missing comma that is not the first one) and it was a bit more involved than I anticipated. I need to regroup and write a generalized version, but I feel like we could land this very specific one in the meantime, after I fix the typo.

**Comment by oli-obk** [comment]
Sgtm

**Comment by estebank** [comment]
Fixed the typo.

**Review by oli-obk — APPROVED** [review]

**Comment by estebank** [comment]
@bors r-

gonna rebase against master, it's probably stale

**Comment by estebank** [comment]
@oli-obk this regressed after my [_other_ change](https://github.com/rust-lang/rust/pull/52649):

```
error: expected token: `,`
  --> <println macros>:4:29
   |
1  | / (  ) => ( print ! ( "\n" ) ) ; ( $ ( $ arg : tt ) * ) => (
2  | | {
3  | | # [ cfg ( not ( stage0 ) ) ] {
4  | | ( $ crate :: io :: _print ( format_args_nl ! ( $ ( $ arg ) * ) ) ) ; } # [
   | |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
5  | | cfg ( stage0 ) ] { print ! ( "{}\n" , format_args ! ( $ ( $ arg ) * ) ) } } )
   | |_____________________________________________________________________________- in this expansion of `println!`
   |
  ::: ../../src/test/ui/macros/missing-comma.rs:12:5
   |
12 |       println!("{}" a);
   |       ----------------- in this macro invocation
```

Closing until I can get back to this.

**Comment by estebank** [comment]
The previous code is still valid for macros other than `println`. [Fixed for `println` too with one line](https://github.com/rust-lang/rust/pull/52397/commits/cce4ea5149e28b938d0af8466dd9af1e489f1b55).

@bors r=oli-obk

**Comment by sunjay** [comment]
This is awesome! Thank you for doing this! I'm the one who filed the original issue about it and I really appreciate that it was fixed. 😁

**Pull request merged**